python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -54,10 +55,44 @@ def __init__(self, instance: "DagsterInstance"):
         # materialization record for a >= cursor, we don't need to query the instance
         self._no_materializations_after_cursor_cache: Dict[AssetKeyPartitionKey, int] = {}
 
+        self._asset_record_cache: Dict[AssetKey, Optional[AssetRecord]] = {}
+
     @property
     def instance(self) -> "DagsterInstance":
         return self._instance
 
+    def prefetch_for_keys(self, asset_keys: Sequence[AssetKey]):
+        """For performance, batches together queries for selected assets"""
+        asset_records = self.instance.get_asset_records(asset_keys)
+        for asset_record in asset_records:
+            self._asset_record_cache[asset_record.asset_entry.asset_key] = asset_record
+
+        # get the latest N materialization records. in most cases, this will hit a large portion of
+        # the latest materialization records per asset key, although a more sophisticated method
+        # of bucketing queries could be used here to guarantee that we actually get the exact set
+        # we're aiming for here.
+        latest_records = self._instance.get_event_records(
```

Comment:
this method of getting relevant event records is a bit suspicious, but my guess is that it'll work out ok in practice. even if all of this work goes to waste (i.e. none of the most recent N asset events are relevant), this is still pretty cheap. When N=2000 (in my test), this query took less than .5s, with the entire prefetch method taking ~1s.